autopush/websocket.py

```diff
@@ -351,55 +390,44 @@ def force_delete(self, result, chid):
                           self.uaid, chid)
         d.addErrback(self.force_delete, chid)
 
-    def process_ack(self, data):
-        updates = data.get("updates")
-        if not updates or not isinstance(updates, list):
+    def ack_update(self, update):
+        chid = update.get("channelID")
+        version = update.get("version")
+        if not chid or not version:
             return
 
-        self.metrics.increment("updates.client.ack")
-        defers = []
-        for update in updates:
-            chid = update.get("channelID")
-            version = update.get("version")
-            if not chid or not version:
-                continue
-
-            skip = False
-            # We always need to delete direct updates
-            if self.direct_updates.get(chid) == version:
-                del self.direct_updates[chid]
-                skip = True
-
-            # If this is the same as a version we sent, delete
-            # as well
-            if self.updates_sent.get(chid) == version:
-                del self.updates_sent[chid]
-            else:
-                # An ack for something we aren't tracking?
-                continue
+        # If its a direct update, remove it and return
+        if self.direct_updates.get(chid) == version:
+            del self.direct_updates[chid]
+            return
 
-            if skip:
-                continue
+        # Remove the update if version matches
+        if self.updates_sent.get(chid) == version:
+            del self.updates_sent[chid]
+        else:
+            return
 
-            # Attempt to delete this notification from storage
-            storage = self.ap_settings.storage
+        # If we ack'd a notification that wasn't direct, delete it
+        d = deferToThread(self.ap_settings.storage.delete_notification,
+                          self.uaid, chid, version)
+        d.addCallback(self.check_ack, self.uaid, chid, version)
```

Comment:
Nope, only an uncaught exception by the function results in the retry. If it fails the conditional check, we move on, since that means it was already deleted or is newer than what we're trying to delete.